### PR TITLE
[Berkeley] Evaluate selectors for non-optional gates

### PR DESCRIPTION
**Pull request by mrmr1993**
This PR adds handling for evaluations of the 'non-optional' gate selectors in pickles, as part of the work to remove the error-prone, expensive, and unwieldy linearization part of kimchi proofs. The proof-systems counterpart can be found in https://github.com/o1-labs/proof-systems/pull/1105.

This PR builds upon #13528.

Checklist:

- [ ] Dependency versions are unchanged
  - Notify Velocity team if dependencies must change in CI
- [ ] Modified the current draft of release notes with details on what is completed or incomplete within this project
- [ ] Document code purpose, how to use it
  - Mention expected invariants, implicit constraints
- [ ] Tests were added for the new behavior
  - Document test purpose, significance of failures
  - Test names should reflect their purpose
- [ ] All tests pass (CI will check this if you didn't)
- [ ] Serialized types are in stable-versioned modules
- [ ] Does this close issues? List them

### CODE REVIEW

**Comment by mrmr1993** [comment]
!ci-build-me

**Comment by mrmr1993** [comment]
!ci-build-me

**Comment by mrmr1993** [comment]
!ci-build-me

**Review by rbonichon — APPROVED** [review]
LGTM

File: src/lib/pickles/step_verifier.ml [hunk 1]
```diff
@@ -599,12 +599,14 @@ struct
              It should be sufficient to fork the sponge after squeezing beta_3 and then to absorb
              the combined inner product.
           *)
-          let num_commitments_without_degree_bound = Nat.N41.n in
+          let num_commitments_without_degree_bound = Nat.N45.n in
```
Comment (dannywillems) [inline src/lib/pickles/step_verifier.ml:602]: That shows an interesting use case of the type encoding features of pickles :)

File: src/lib/pickles_types/plonk_types.ml [hunk 2]
```diff
@@ -586,9 +697,27 @@ module Evals = struct
   *)
 
   let to_list
-      { w; coefficients; z; s; generic_selector; poseidon_selector; lookup } =
+      { w
```
Comment (dannywillems) [inline src/lib/pickles_types/plonk_types.ml:700]: A bit picky, but the comment above it out-dated with these changes.

File: src/lib/pickles_types/plonk_types.mli [hunk 3]
```diff
@@ -247,6 +251,10 @@ module Evals : sig
     ; s : 'a Permuts_minus_1_vec.t
     ; generic_selector : 'a
     ; poseidon_selector : 'a
+    ; complete_add_selector : 'a
```
Comment (dannywillems) [inline src/lib/pickles_types/plonk_types.mli:254]: Also picky, but accessors are added for the other selectors. Do we want to stay consistent?

**Comment by dannywillems** [comment]
What is the impact of
```
  Now: type 'a t = { w: 'a Columns_vec.Stable.V1.t ; coefficients: 'a Columns_vec.Stable.V1.t ; z: 'a ; s: 'a Permuts_minus_1_vec.Stable.V1.t ; generic_selector: 'a ; poseidon_selector: 'a ; complete_add_selector: 'a ; mul_selector: 'a ; emul_selector: 'a ; endomul_scalar_selector: 'a ; lookup: 'a Lookup.Stable.V1.t option }[@@deriving bin_io]
```, see CI, job `version compatibility linter`.

**Comment by dannywillems** [comment]
I checked by grepping `poseidon_selector` if the new gates selectors were added where poseidon_selector was. LGTM, except the comments above. I also rely on the type encoding of vectors for selectors to be sure selectors are added where necessary in pickles.
The changes in snarkyjs-bindings seems to be only adding the selectors' evaluation in `ProofEvaluations` in the `plonk_proof.rs`. No other places in `bindings.js` or `wasm/plonk_proof.rs`, `poseidon_selector` is where new gates are not.

**Review by querolita — APPROVED** [review]

**Comment by rbonichon** [comment]
!ci-build-me

**Comment by rbonichon** [comment]
!ci-build-me

**Comment by joseandro** [comment]
!ci-build-me

**Comment by rbonichon** [comment]
!ci-build-me

**Comment by rbonichon** [comment]
!ci-build-me

**Comment by dannywillems** [comment]
!ci-build-me

**Comment by dannywillems** [comment]
!ci-build-me

**Comment by mrmr1993** [comment]
!approved-for-mainnet

**Pull request merged**